File: Tests/CppWinRTTests/UWP/Cpp/SocialManagerIntegration.cpp
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//
// Copyright (c) Microsoft Corporation. All rights reserved
#include "pch.h"
#include "GameLogic\Game.h"
#include <time.h>
#include "Utils\Utils.h"
#include "Utils\PerformanceCounters.h"

using namespace Concurrency;
using namespace Windows::Foundation::Collections;
using namespace Windows::Foundation;
using namespace xbox::services;
using namespace xbox::services::social::manager;
using namespace Sample;

void
Game::InitializeSocialManager()
{
    m_socialManager = social_manager::get_singleton_instance();
}

void
Game::AddUserToSocialManager(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user
    )
{
    {
        std::lock_guard<std::mutex> guard(m_socialManagerLock);

        stringstream_t source;
        source << _T("Adding user ");
        source << user->gamertag();
        source << _T(" to SocialManager");
        Log(source.str());

        m_socialManager->add_local_user(user, social_manager_extra_detail_level::preferred_color_level | social_manager_extra_detail_level::title_history_level);
    }

    CreateSocialGroupsBasedOnUI(user);
}

void
Game::RemoveUserFromSocialManager(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user
    )
{
    std::lock_guard<std::mutex> guard(m_socialManagerLock);

    stringstream_t source;
    source << _T("Removing user ");
    source << user->gamertag();
    source << _T(" from SocialManager");
    Log(source.str());

    auto it = m_socialGroups.begin();
    while (it != m_socialGroups.end()) 
    {
        std::shared_ptr<xbox::services::social::manager::xbox_social_user_group> group = *it;
        if (wcscmp(group->local_user()->xbox_user_id().c_str(), user->xbox_user_id().c_str()))
        {
            it = m_socialGroups.erase(it);
        }
        else 
        {
            ++it;
        }
    }

    m_socialManager->remove_local_user(user);
}

void
Game::CreateOrUpdateSocialGroupFromList(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user,
    _In_ std::vector<string_t> xuidList
    )
{
    for (auto group : m_socialGroups)
    {
        if (group->social_user_group_type() == social_user_group_type::user_list_type && group->local_user()->xbox_user_id() == user->xbox_user_id())
        {
            m_socialManager->update_social_user_group(group, xuidList);
            return;
        }
    }

    if( xuidList.size() > 0 )
    {
        auto result = m_socialManager->create_social_user_group_from_list(user, xuidList);
        if (!result.err())
        {
            std::lock_guard<std::mutex> guard(m_socialManagerLock);
            m_socialGroups.push_back(result.payload());
        }
    }
}

void
Game::CreateSocialGroupFromFilters(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user,
    _In_ presence_filter presenceFilter,
    _In_ relationship_filter relationshipFilter
    )
{
    auto result = m_socialManager->create_social_user_group_from_filters(user, presenceFilter, relationshipFilter);
    if (!result.err())
    {
        std::lock_guard<std::mutex> guard(m_socialManagerLock);
        m_socialGroups.push_back(result.payload());
    }
}

void
Game::DestorySocialGroupFromList(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user
    )
{
    std::lock_guard<std::mutex> guard(m_socialManagerLock);

    auto it = m_socialGroups.begin();
    while (it != m_socialGroups.end())
    {
        std::shared_ptr<xbox::services::social::manager::xbox_social_user_group> group = *it;
        if (wcscmp(group->local_user()->xbox_user_id().c_str(), user->xbox_user_id().c_str()) == 0 &&
            group->social_user_group_type() == social_user_group_type::user_list_type)
        {
            m_socialManager->destroy_social_user_group(group);
            it = m_socialGroups.erase(it);
        }
        else
        {
            ++it;
        }
    }
}

void
Game::DestroySocialGroup(
    _In_ std::shared_ptr<xbox::services::system::xbox_live_user> user,
    _In_ presence_filter presenceFilter,
    _In_ relationship_filter relationshipFilter
    )
{
    std::lock_guard<std::mutex> guard(m_socialManagerLock);

    auto it = m_socialGroups.begin();
    while (it != m_socialGroups.end())
    {
        std::shared_ptr<xbox::services::social::manager::xbox_social_user_group> group = *it;
        if (wcscmp(group->local_user()->xbox_user_id().c_str(), user->xbox_user_id().c_str()) == 0 &&
            group->presence_filter_of_group() == presenceFilter && 
            group->relationship_filter_of_group() == relationshipFilter)
        {
            m_socialManager->destroy_social_user_group(group);
            it = m_socialGroups.erase(it);
        }
        else
        {
            ++it;
        }
    }
}

void
Game::UpdateSocialManager()
{
    std::lock_guard<std::mutex> guard(m_socialManagerLock);

#if PERF_COUNTERS
    auto perfInstance = performance_counters::get_singleton_instance();
    perfInstance->begin_capture(L"no_updates");
    perfInstance->begin_capture(L"updates");
#endif
    auto socialEvents = m_socialManager->do_work();
#if PERF_COUNTERS
    if (socialEvents.empty())
    {
        perfInstance->end_capture(L"no_updates");
    }
    else
    {
        perfInstance->end_capture(L"updates");
    }
#endif

    LogSocialEventList(socialEvents);
}

std::vector<std::shared_ptr<xbox::services::social::manager::xbox_social_user_group>> Game::GetSocialGroups()
{
    std::lock_guard<std::mutex> guard(m_socialManagerLock);
    return m_socialGroups;
}

File: Tests/CSharpTestApp/UWP/SocialManagerIntegration.cs
using System;
using System.Collections.Generic;
using Microsoft.Xbox.Services;
using Microsoft.Xbox.Services.System;
using Microsoft.Xbox.Services.Social.Manager;

public class SocialManagerIntegration
{
    Social.MainPage m_ui;
    SocialManager m_socialManager = SocialManager.SingletonInstance;
    List<XboxSocialUserGroup> m_socialUserGroups = new List<XboxSocialUserGroup>();
    string m_logText = string.Empty;
    List<string> m_logLines = new List<string>();

    public void Init(Social.MainPage ui)
    {
        m_ui = ui;
    }

    public void AddUser(XboxLiveUser user)
    {
        lock (m_socialManager)
        {
            m_ui.LogEvent(string.Format("Adding user {0} to SocialManager", user.Gamertag));
            m_socialManager.AddLocalUser(user, SocialManagerExtraDetailLevel.NoExtraDetail);
            m_ui.CreateSocialGroupsBasedOnUI(user);
        }
    }

    public void CreateOrUpdateSocialGroupFromList(XboxLiveUser user, List<string> xuidList)
    {

        foreach (XboxSocialUserGroup userGroup in m_socialUserGroups)
        {
            if (userGroup.LocalUser.XboxUserId == user.XboxUserId && userGroup.SocialUserGroupType == SocialUserGroupType.UserListType)
            {
                m_socialManager.UpdateSocialUserGroup(userGroup, xuidList);
                return;
            }
        }
        if ( xuidList.Count > 0 )
        {
            XboxSocialUserGroup socialUserGroup = m_socialManager.CreateSocialUserGroupFromList(user, xuidList);
            lock (m_socialManager)
            {
                m_socialUserGroups.Add(socialUserGroup);
            }
        }
    }

    public void DestorySocialGroupFromList(XboxLiveUser user)
    {
        lock (m_socialManager)
        {
            foreach (XboxSocialUserGroup socialUserGroup in m_socialUserGroups)
            {
                if (socialUserGroup.LocalUser.XboxUserId == user.XboxUserId &&
                    socialUserGroup.SocialUserGroupType == SocialUserGroupType.UserListType)
                {
                    m_socialUserGroups.Remove(socialUserGroup);
                    m_socialManager.DestroySocialUserGroup(socialUserGroup);
                    break;
                }
            }
        }
    }

    public void CreateSocialGroupFromFilters(XboxLiveUser user, PresenceFilter presenceFilter, RelationshipFilter relationshipFilter)
    {
        XboxSocialUserGroup socialUserGroup = m_socialManager.CreateSocialUserGroupFromFilters(user, presenceFilter, relationshipFilter);
        lock (m_socialManager)
        {
            m_socialUserGroups.Add(socialUserGroup);
        }
    }

    public void DestroySocialGroup(XboxLiveUser user, PresenceFilter presenceFilter, RelationshipFilter relationshipFilter)
    {
        lock (m_socialManager)
        {
            foreach (XboxSocialUserGroup socialUserGroup in m_socialUserGroups)
            {
                if( socialUserGroup.LocalUser.XboxUserId == user.XboxUserId &&
                    socialUserGroup.PresenceFilterOfGroup == presenceFilter && 
                    socialUserGroup.RelationshipFilterOfGroup == relationshipFilter )
                {
                    m_socialUserGroups.Remove(socialUserGroup);
                    m_socialManager.DestroySocialUserGroup(socialUserGroup);
                    break;
                }
            }
        }
    }

    public void RemoveUser(XboxLiveUser user)
    {
        lock (m_socialManager)
        {
            List<XboxSocialUserGroup> socialUserGroups = new List<XboxSocialUserGroup>(m_socialUserGroups);
            foreach(XboxSocialUserGroup socialUserGroup in socialUserGroups)
            {
                if( socialUserGroup.LocalUser.XboxUserId == user.XboxUserId )
                {
                    m_socialUserGroups.Remove(socialUserGroup);
                }
            }

            m_ui.LogEvent(string.Format("Removing user {0} from SocialManager", user.Gamertag));
            m_socialManager.RemoveLocalUser(user);
        }
    }

    public void Update()
    {
        List<XboxSocialUserGroup> socialUserGroups = null;
        IReadOnlyList<SocialEvent> socialEventList = null;
        lock (m_socialManager)
        {
            socialUserGroups = m_socialUserGroups;
            socialEventList = m_socialManager.DoWork();
        }
        RenderSocialGroupList(socialUserGroups);
        LogSocialEventList(socialEventList);
    }

    void LogSocialEventList(IReadOnlyList<SocialEvent> socialEventList)
    {
        foreach (SocialEvent socialEvent in socialEventList)
        {
            if ( socialEvent.ErrorCode != 0 )
            {
                m_ui.LogEvent(string.Format("Event: {0} ErrorCode: 0x{1:X} ErrorMessage: {2}", 
                    socialEvent.EventType.ToString(), 
                    socialEvent.ErrorCode,
                    socialEvent.ErrorMessage));
            }
            else
            {
                string usersAffected = string.Empty;
                foreach( String u in socialEvent.UsersAffected )
                {
                    usersAffected += u;
                    usersAffected += ", ";
                }
                m_ui.LogEvent(string.Format("Event: {0} UserAffected: {1}", socialEvent.EventType.ToString(), usersAffected));
            }
        }
    }

    async void RenderSocialGroupList(List<XboxSocialUserGroup> socialUserGroups)
    {
        Windows.UI.Core.CoreDispatcher UIDispatcher = Windows.UI.Xaml.Window.Current.CoreWindow.Dispatcher;
        await UIDispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal,
            () =>
            {
                m_ui.ClearSocialGroup();

                foreach (XboxSocialUserGroup socialUserGroup in socialUserGroups)
                {
                    m_ui.AppendToSocialGroup("--------------------");
                    if( socialUserGroup.SocialUserGroupType == SocialUserGroupType.FilterType )
                    {
                        m_ui.AppendToSocialGroup(
                            string.Format("Group from filter: {0} {1}", socialUserGroup.PresenceFilterOfGroup.ToString(), socialUserGroup.RelationshipFilterOfGroup.ToString())
                            );
                    }
                    else
                    {
                        m_ui.AppendToSocialGroup("Group from custom list");
                    }

                    IReadOnlyList<XboxSocialUser> userList = socialUserGroup.Users;
                    foreach (XboxSocialUser socialUser in userList)
                    {
                        m_ui.AppendToSocialGroup("Gamertag: " + socialUser.Gamertag + ". Status: " + socialUser.PresenceRecord.UserState.ToString());
                    }

                    if (userList.Count == 0)
                    {
                        m_ui.AppendToSocialGroup("No friends found");
                    }
                }
            });
    }
}
